forge-game/src/main/java/forge/game/ability/effects/SeekEffect.java

```diff
@@ -0,0 +1,94 @@
+import forge.game.ability.AbilityKey;
+import forge.game.ability.AbilityUtils;
+import forge.game.ability.SpellAbilityEffect;
+import forge.game.card.*;
+import forge.game.player.Player;
+import forge.game.spellability.SpellAbility;
+import forge.game.trigger.TriggerType;
+import forge.game.zone.ZoneType;
+import forge.util.Aggregates;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+public class SeekEffect extends SpellAbilityEffect {
+
+    /* (non-Javadoc)
+     * @see forge.game.ability.SpellAbilityEffect#getStackDescription(forge.game.spellability.SpellAbility)
+     */
+    @Override
+    protected String getStackDescription(SpellAbility sa) {
+        return sa.getDescription();
+    }
+
+    @Override
+    public void resolve(SpellAbility sa) {
+        final Card source = sa.getHostCard();
+        final Game game = source.getGame();
+
+        List<String> seekTypes = Lists.newArrayList();
+        if (sa.hasParam("Types")) {
+            seekTypes.addAll(Arrays.asList(sa.getParam("Types").split(",")));
+        } else {
+            seekTypes.add(sa.getParamOrDefault("Type", "Card"));
+        }
+
+        int seekNum = AbilityUtils.calculateAmount(source, sa.getParamOrDefault("Num", "1"), sa);
+        if (seekNum <= 0) {
+            return;
+        }
+
+        final CardZoneTable triggerList = new CardZoneTable();
+        CardCollectionView lastStateBattlefield = game.copyLastStateBattlefield();
+        CardCollectionView lastStateGraveyard = game.copyLastStateGraveyard();
+
+        for (Player seeker : getTargetPlayers(sa)) {
+            if (!seeker.isInGame()) {
+                continue;
+            }
+
+            CardCollection soughtCards = new CardCollection();
+
+            for (String seekType : seekTypes) {
+                CardCollection pool;
+                if (sa.hasParam("DefinedCards")) {
+                    pool = AbilityUtils.getDefinedCards(source, sa.getParam("DefinedCards"), sa);
+                } else {
+                    pool = new CardCollection(seeker.getCardsIn(ZoneType.Library));
+                }
+                if (!seekType.equals("Card")) {
+                    pool = CardLists.getValidCards(pool, seekType, source.getController(), source, sa);
+                }
+                if (pool.isEmpty()) {
+                    continue; // can't find if nothing to seek
+                }
+
+                for (final Card c : Aggregates.random(pool, seekNum)) {
+
+                    Map<AbilityKey, Object> moveParams = AbilityKey.newMap();
+                    moveParams.put(AbilityKey.LastStateBattlefield, lastStateBattlefield);
+                    moveParams.put(AbilityKey.LastStateGraveyard, lastStateGraveyard);
+                    Card movedCard = game.getAction().moveToHand(c, sa, moveParams);
+                    soughtCards.add(movedCard);
+
+                    triggerList.put(ZoneType.Library, movedCard.getZone().getZoneType(), movedCard);
```

Comment:
what if in a Commander Game, the commander is put into libary (and the player kept it there) and then it got seeked but THEN the player choose to put the Commander into Command Zone instead? then it shouldn't be put to hand?

>903.9b If a commander would be put into its owner’s hand or library from anywhere, its owner may put it into the command zone instead. This replacement effect may apply more than once to the same event. This is an exception to rule 614.5.

Reply:
@Northmoc you need to take care if the card is not moved to hand because of some kind of replacement effect